torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -54,6 +56,70 @@ def reset_table(self) -> None:
 kernel_side_table = KernelSideTable()
 
 
+###############################################################################
+# Mutation Tracker
+
+
+@dataclasses.dataclass
+class MutationInfo:
+    mutated: bool = False
+    used_in_unknown: bool = False
+
+
+# Super basic mutation tracking pass that tracks which inputs are used in stores
+# It bails if any of the inputs are used in non tl.load/tl.store positions.
+class MutationTracker(ast.NodeVisitor):
+    def __init__(self, infos) -> None:
+        super().__init__()
+        self.infos = infos
+        self.in_load = False
+        self.in_store = False
+
+    def visit_Name(self, node):
+        if node.id not in self.infos:
+            return
+        if self.in_load:
+            pass
+        elif self.in_store:
+            self.infos[node.id].mutated = True
+        else:
+            self.infos[node.id].used_in_unknown = True
+
+    def visit_Call(self, node):
+        if (
+            isinstance(node.func, ast.Attribute)
+            and isinstance(node.func.value, ast.Name)
+            and node.func.value.id == "tl"
+        ):
+            if node.func.attr == "store":
+                assert self.in_store is False
+                self.in_store = True
+                self.generic_visit(node)
```

Comment:
IIUC, `self.generic_visit` visits *all* children of the `tl.store` node from here. I think, we should be more specific than just saying "we're in store and every input covered here is mutated". This is because `tl.store` has [multiple parameters](https://triton-lang.org/main/python-api/generated/triton.language.store.html#triton.language.store) and we only care about the inputs covered by the first one: `pointer`. In other words, if any inputs are used only in the `value` and / or `mask` computation, those are (generally) not mutated.

It may be that the code path from those additional argument back to the non-modified inputs always goes through `tl.load`, though, in which case `if self.in_load: pass` will happen above. But not quite sure about this.

Reply:
Keyword args and regular args get parsed differently, so doing it only for first arg was annoying. I doubt anyone will use input tensor in other arguments load so this was simple enough?

Reply:
So the current code does analysis on all args of `tl.store`, both keyword and regular? Using read-only input tensor in the dependency tree of other arguments is possible. E.g., the values loaded from read-only arg are used in the (second) `value` argument of `tl.store`. But, as far as I can think, there will always be a `tl.load` somewhere on the code path from the other (not `pointer`) arguments to the read-only inputs. So, if you think it is non-trivial to consider only the first argument of `tl.store`, it sounds fine to start with this version and see if we run into edge cases later.